src/dxss/gmres.py

```diff
@@ -23,20 +23,28 @@ def __init__(
         callback_sol: Optional[Callable[[PETSc.Vec], None]] = None,
         printrates: bool = False,
     ):
+        self.mat = mat
         if atol is None and tol is None:
             tol = 1e-12
-        self.mat = mat
         self.pre = pre
         self.tol = tol
         self.atol = atol
         self.maxiter = maxiter
         self.callback = callback
         self.callback_sol = callback_sol
         self.printrates = printrates
-        self.residuals = []
-        self.iterations = 0
 
-    def Solve(
+        self.residuals: list[float] = []
+        self.iterations: int = 0
+
+    def _solve_impl(self, rhs: PETSc.Vec, sol: PETSc.Vec) -> None:
+        """
+        Method-specific solving function called by `LinearSolver.solve`.
+
+        This is a no-op, and is intended for derived classes to override.
+        """
```

Comment:
```suggestion
        This is a no-op, and is intended for derived classes to override.
        """
        raise NotImplementedError
```

Possibly out of scope but more directly enforces what the comment is saying (unless there are cases where we want to use no-op default)?

Reply:
Nice, thanks.

I was thinking of making it an [`@abstractmethod`](https://docs.python.org/3/library/abc.html#abc.abstractmethod) but perhaps that's more for future refactorings.